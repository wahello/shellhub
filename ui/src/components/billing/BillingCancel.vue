<template>
  <fragment>
    <v-btn
      color="red darken-1"
      outlined
      data-test="cancel-btn"
      @click="dialog = !dialog"
    >
      Cancel
    </v-btn>

    <v-dialog
      v-model="dialog"
      max-width="510"
    >
      <v-card data-test="billingWarning-dialog">
        <v-card-title class="headline grey lighten-2 text-center">
          Are you sure?
        </v-card-title>

        <v-card-text class="mt-4 mb-3 pb-1">
          Canceling the subscription will generate an invoice,
          estimated <b> {{ nextPaymentDue | formatCurrency }} </b> for the time of use.
        </v-card-text>

        <v-card-actions>
          <v-spacer />

          <v-btn
            text
            data-test="close-btn"
            @click="dialog=!dialog"
          >
            Close
          </v-btn>

          <v-btn
            text
            data-test="cancel-btn"
            @click="cancelSubscription()"
          >
            Cancel
          </v-btn>
        </v-card-actions>
      </v-card>
    </v-dialog>
  </fragment>
</template>

<script>

import formatCurrency from '@/components/filter/currency';

export default {
  name: 'BillingWarning',

  filters: { formatCurrency },

  props: {
    nextPaymentDue: {
      type: Number,
      required: true,
    },
  },

  data() {
    return {
      dialog: false,
    };
  },

  methods: {
    async cancelSubscription() {
      try {
        await this.$store.dispatch('billing/cancelSubscription');
        this.$store.dispatch('snackbar/showSnackbarSuccessAction', this.$success.cancelSubscription);

        this.$emit('update');
        this.$store.dispatch('devices/setDeviceWarning', this.$store.getters['stats/stats'].registered_devices > 3);
      } catch {
        this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.cancelSubscription);
      }
    },
  },
};

</script>
