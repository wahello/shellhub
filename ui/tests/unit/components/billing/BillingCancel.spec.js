import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import BillingCancel from '@/components/billing/BillingCancel';

describe('BillingCancel', () => {
  const localVue = createLocalVue();
  const vuetify = new Vuetify();
  localVue.use(Vuex);

  document.body.setAttribute('data-app', true);

  let wrapper;

  const nextPaymentDue = 1234;

  const stats = {
    registered_devices: 0,
    online_devices: 0,
    active_sessions: 0,
    pending_devices: 1,
    rejected_devices: 0,
  };

  const store = new Vuex.Store({
    namespaced: true,
    state: {
      stats,
    },
    getters: {
      'stats/stats': (state) => state.stats,
    },
    actions: {
      'billing/cancelSubscription': () => {},
      'devices/setDeviceWarning': () => {},
      'snackbar/showSnackbarSuccessAction': () => {},
      'snackbar/showSnackbarErrorAction': () => {},
    },
  });

  ///////
  // In this case, focus of the test is button rendering.
  ///////

  describe('Button', () => {
    beforeEach(() => {
      wrapper = mount(BillingCancel, {
        store,
        localVue,
        stubs: ['fragment'],
        propsData: { nextPaymentDue },
        vuetify,
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Receive data in props', () => {
      expect(wrapper.vm.nextPaymentDue).toBe(nextPaymentDue);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', () => {
      expect(wrapper.find('[data-test="cancel-btn"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="billingWarning-dialog"]').exists()).toBe(false);
    });
  });

  ///////
  // In this case, focus of the test is dialog rendering.
  ///////

  describe('Dialog', () => {
    beforeEach(() => {
      wrapper = mount(BillingCancel, {
        store,
        localVue,
        stubs: ['fragment'],
        propsData: { nextPaymentDue },
        vuetify,
      });

      wrapper.setData({ dialog: true });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Receive data in props', () => {
      expect(wrapper.vm.nextPaymentDue).toBe(nextPaymentDue);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', () => {
      expect(wrapper.find('[data-test="cancel-btn"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="billingWarning-dialog"]').exists()).toBe(true);
    });
  });
});
