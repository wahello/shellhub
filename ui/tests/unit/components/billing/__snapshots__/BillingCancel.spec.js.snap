// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BillingCancel Button Renders the component 1`] = `
<fragment-stub><button type="button" class="v-btn v-btn--outlined theme--light v-size--default red--text text--darken-1" data-test="cancel-btn"><span class="v-btn__content">
    Cancel
  </span></button>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;

exports[`BillingCancel Dialog Renders the component 1`] = `
<fragment-stub><button type="button" class="v-btn v-btn--outlined theme--light v-size--default red--text text--darken-1" data-test="cancel-btn"><span class="v-btn__content">
    Cancel
  </span></button>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;
